api/src/middleware/auth.ts

```diff
@@ -0,0 +1,17 @@
+import jwt from "jsonwebtoken";
+import env from "../../environment";
+
+export default function auth(req, res, next) {
+    const authHeader = req.header("authorization");
+    if (!authHeader) {
+        return res.status(401).json({ msg: "No token, authorization denied." });
+    }
+    const token = authHeader.split(" ")[1];
+    jwt.verify(token, env.ACCESS_TOKEN_SECRET, (err, user_token) => {
+        if (err) {
+            return res.status(403).json({ msg: "Token is not valid." });
+        }
+        req.user = user_token;
+        next();
```

Comment:
Só pra ver se eu entendi:
em um eventual front, a requisição é feita com o access token no header de 'Authorization'. Se a requisição der certo, tudo bem. Se ela falhar, o front vai fazer uma requisição para dar refresh no token de acesso e, se der certo, ele faz a requisição inicial de novo?

Eu não entendo muito da segurança envolvida, mas não seria mais simples se a requisição de refresh fosse feita aqui mesmo, no middleware?

Reply:
Entendeu certo!
Mas não dá pra fazer isso que você sugeriu, porque middleware tem que ter o seguinte comportamento:
- Se tiver algum problema, retornar erro
- Se não, chamar a próxima função pra ela retornar alguma coisa

O problema é que a rota de refresh token necessariamente precisa retornar um access token. Não adianta só gerar um token novo, o front precisa pegar esse token e armazenar em memória pra poder mandar como header de outras rotas.
Seria pior se o front mandasse uma request pra, digamos, listar uma wishlist do usuário, recebesse um status 200 de sucesso, mas o retorno é um token ao invés da resposta que ele queria.

Sei que é estranho, mas é assim mesmo que funciona essa dinâmica de refresh e access token. Dá uma olhada nessa página do Medium, ela explica como um front-end tem que ser montado pra funcionar o refresh token:
https://medium.com/swlh/how-to-implement-refresh-token-functionality-front-end-eff58ce52564

Reply:
No login, o front armazena o access token e quando ele expira. Quando ele vai fazer uma request, se o token expirou, ele primeiro chama o refresh token pra armazenar o novo access token, e depois sim ele faz o que ele queria

Reply:
Tirei esses conceitos inicialmente desse tutorial aqui: https://www.youtube.com/watch?v=favjC6EKFgw

Reply:
Ahh, entendi. Obrigado por explicar, faz mais sentido agora